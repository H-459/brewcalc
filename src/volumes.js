// @flow
import { sum, options } from './utils.js'
import { FermentableTypes } from './types/fermentable'
import type { Fermentable } from './types/fermentable'
import { MashType } from './types/mashStep'
import type { Recipe } from './types/recipe'
import type { Equipment } from './types/equipment'

export const calculateVolumes = (
  { fermentables }: Recipe,
  equipment: Equipment
) => {
  const mashGrainWeight = sum(
    fermentables.map(
      ({ amount, type }) => type == FermentableTypes.grain ? amount : 0
    )
  )
  return {
    //===
    //Mashing
    //===
    //TotalWater
    mashGrainWeight
    //GrainAbsorption
    //MashTunAddition
    //TotalMashWaterAdds
    //TunDeadspace
    //MashTunVolume
    //MashTunVolumeNeeded
    //WaterAvailFromMash
    //SpargeVol
    //===
    //Boiling
    //===
    //KettleTopUp
    //EstPreBoilVolume
    //MeasPreBoilVolume
    //EvapRate
    //BoilOffVolume
    //PostBoilVolume
    //CoolingShrinkage
    //CoolLossPrc
    //TrubLoss
    //===
    //Fermenting
    //===
    //FermentationTopUp
    //BatchSize
    //MeasBatchSize
    //StarterSize
    //FermentationLoss
    //EstBottlingVol
    //MeasBottlingVol
    //*/
  }
}
/*
export const totalGrains = recipe => {
  return sum(
    recipe.fermentables.map(
      ({ amount, addAfterBoil }) => !addAfterBoil ? amount : 0
    )
  )
}

export const totalMashGrains = recipe => {
  return sum(
    recipe.fermentables.map(({ amount, type }) => {
      if (type == FermentableTypes.grain) return amount
      else return 0
    })
  )
}

export const totalMashWater = recipe => {
  var totalWater = 0

  if (recipe.mash.mashSteps == null || recipe.mash.mashSteps == 'undefined')
    return 0

  recipe.mash.mashSteps.map(mashStep => {
    if (mashStep.type != MashType.decoction) {
      totalWater += mashStep.infuseAmount
    }
  })

  return totalWater + recipe.equipment.lauterDeadspace
}

export const vaterAvailableFromMash = recipe => {
  return totalMashWater(recipe) -
    totalMashGrains(recipe) * options().grainAbsorb
}

export const spargeVol = recipe => {
  const grainAbsLoss = options().grainAbsorb * totalMashGrains(recipe)

  const vol = recipe.equipment.boilSize +
    grainAbsLoss -
    recipe.equipment.topUpKettle +
    options().mashTunDeadSpace -
    totalMashWater(recipe)
  return vol
}

export const calcWater = (
  {
    batchSize,
    totalGrains,
    boilTime,
    evapRate,
    coolingLossPct,
    mashTunDeadSpace = 0,
    kettleDeadSpace = 0,
    dissolvedVol = 0,
    hopWaterLoss = 0,
    pumpWaterLoss = 0
  }
) => {
  const coolingLoss = batchSize * coolingLossPct

  const grainAbsLoss = options().grainAbsorb * totalGrains

  const boilOffLoss = evapRate * boilTime / 60

  const intoFermenterVol = batchSize
  const fromKettleVol = intoFermenterVol + pumpWaterLoss
  const postBoilVol = fromKettleVol + hopWaterLoss + kettleDeadSpace
  const hotPostBoilVol = postBoilVol + coolingLoss
  const preBoilVol = hotPostBoilVol + boilOffLoss
  const preBoilWaterVol = preBoilVol - dissolvedVol
  const totalWater = preBoilWaterVol + grainAbsLoss + mashTunDeadSpace

  return {
    grainAbsLoss,
    totalWater,
    preBoilWaterVol,
    preBoilVol,
    hotPostBoilVol,
    postBoilVol,
    fromKettleVol
  }
}
*/
