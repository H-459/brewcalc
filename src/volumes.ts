import { sum } from './utils'
import {
  VolumeType,
  MassType,
  EquipmentItemType,
  BoilProcedureType,
  MashStepType,
} from '../types/beerjson'
import { convertMeasurableValue } from './units'

const defaultVolume: VolumeType = {
  value: 0,
  unit: 'gal',
}

const defaultBoil: BoilProcedureType = {
  pre_boil_size: {
    value: 0,
    unit: 'gal',
  },
  boil_time: {
    value: 0,
    unit: 'min',
  },
}

const coolingShrinkageRate = 0.04

const convertToGallons = (volume: VolumeType) =>
  convertMeasurableValue(volume, 'gal')

// 0.96 - number of fl. ounces of water absorbed per ounce of the grain
// 128 fl. ounces in gallon, 16 ounces in pound
const grainAbsorptionRatio = (0.96 / 128) * 16

const calcGrainAbsorption = (grainWeight: MassType): VolumeType => {
  const value = convertMeasurableValue(grainWeight, 'lb') * grainAbsorptionRatio
  return {
    value,
    unit: 'gal',
  }
}

const calcMashWaterVolume = (
  mash_steps: Array<MashStepType> = []
): VolumeType => {
  const value = sum(
    mash_steps.map(({ type, amount }: MashStepType) =>
      type === 'infusion' ? convertToGallons(amount) : 0
    )
  )
  return {
    value,
    unit: 'gal',
  }
}

export const calcMashVolumes = (
  pre_boil_size: VolumeType,
  mashSteps: Array<MashStepType>,
  mashGrainWeight: MassType,
  equipment: {
    hlt?: EquipmentItemType
    mash_tun?: EquipmentItemType
    brew_kettle?: EquipmentItemType
    fermenter?: EquipmentItemType
  }
): {
  mash_volume: VolumeType
  sparge_volume: VolumeType
  total_volume: VolumeType
} => {
  const mashWaterVolume = calcMashWaterVolume(mashSteps)

  const grainAbsorption = calcGrainAbsorption(mashGrainWeight)

  const mashLoss =
    equipment.mash_tun != null ? convertToGallons(equipment.mash_tun.loss) : 0

  const spargeVolumeValue =
    convertToGallons(pre_boil_size) +
    grainAbsorption.value -
    mashWaterVolume.value +
    mashLoss

  const spargeVolume: VolumeType = {
    value: spargeVolumeValue,
    unit: 'gal',
  }

  const totalVolume: VolumeType = {
    value: mashWaterVolume.value + spargeVolume.value,
    unit: 'gal',
  }

  return {
    mash_volume: mashWaterVolume,
    sparge_volume: spargeVolume,
    total_volume: totalVolume,
  }
}

export const calcBoilVolumes = (
  batch_size: VolumeType,
  boil: BoilProcedureType = defaultBoil,
  equipment: {
    hlt?: EquipmentItemType
    mash_tun?: EquipmentItemType
    brew_kettle?: EquipmentItemType
    fermenter?: EquipmentItemType
  }
): { pre_boil_size: VolumeType } => {
  const boilProfile = boil || defaultBoil

  const postBoilVolume = convertToGallons(batch_size)
  const boilLoss =
    equipment.brew_kettle != null
      ? convertToGallons(equipment.brew_kettle.loss)
      : defaultVolume.value
  const boilRate =
    equipment.brew_kettle != null &&
    equipment.brew_kettle.boil_rate_per_hour != null
      ? convertToGallons(equipment.brew_kettle.boil_rate_per_hour)
      : 0
  const boilOffVolume = (boilRate * boilProfile.boil_time.value) / 60
  const coolingShrinkage = postBoilVolume * coolingShrinkageRate
  const preBoilVolume =
    postBoilVolume + boilOffVolume + boilLoss + coolingShrinkage

  return {
    pre_boil_size: {
      value: preBoilVolume,
      unit: 'gal',
    },
  }
}
