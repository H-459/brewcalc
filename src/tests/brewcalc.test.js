// @flow
declare var test: any;
declare var expect: any;

import {
  originalGravity,
  originalGravityPoints,
  finalGravity,
  finalGravityPoints,
  boilGravity
} from '../brewcalc'

import { calculateVolumes } from '../volumes'

import { recipe as AussieAle } from './data/recipe/AussieAle.js'
import { recipe as MiddyPig } from './data/recipe/Middy Pig.js'

test('originalGravity', () => {
  const ogPtsAA = originalGravityPoints(AussieAle, AussieAle.equipment)
  const ogPtsMP = originalGravityPoints(MiddyPig, MiddyPig.equipment)

  expect(originalGravity(AussieAle.batchSize, ogPtsAA)).toBeCloseTo(
    AussieAle.og,
    3
  )
  expect(originalGravity(MiddyPig.batchSize, ogPtsMP)).toBeCloseTo(
    MiddyPig.og,
    3
  )
})

test('finalGravity', () => {
  const fgPtsAA = finalGravityPoints(AussieAle, AussieAle.equipment)
  expect(finalGravity(AussieAle.batchSize, fgPtsAA)).toBeCloseTo(
    AussieAle.fg,
    2
  )

  const fgPtsMP = finalGravityPoints(MiddyPig, MiddyPig.equipment)
  expect(finalGravity(MiddyPig.batchSize, fgPtsMP)).toBeCloseTo(MiddyPig.fg, 2)
})

test('boilGravity', () => {
  const ogAussieAle = originalGravity(
    AussieAle.batchSize,
    originalGravityPoints(AussieAle, AussieAle.equipment)
  )
  expect(
    boilGravity(AussieAle.batchSize, AussieAle.boilSize, ogAussieAle)
  ).toBeCloseTo(1.031, 2)

  const ogMiddyPig = originalGravity(
    MiddyPig.batchSize,
    originalGravityPoints(MiddyPig, MiddyPig.equipment)
  )
  expect(
    boilGravity(MiddyPig.batchSize, MiddyPig.boilSize, ogMiddyPig)
  ).toBeCloseTo(1.084, 2)
})

test('mashGrainWeight', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).mashGrainWeight
  ).toBeCloseTo(5.26, 2)
})
test('grainAbsorption', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).grainAbsorption
  ).toBeCloseTo(0.66, 2)
})

test('totalMashWaterAdds', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).totalMashWaterAdds
  ).toBeCloseTo(24.45, 2)
})

test('mashVolumeNeeded', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).mashVolumeNeeded
  ).toBeCloseTo(27.97, 2)
})

test('waterAvailFromMash', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).waterAvailFromMash
  ).toBeCloseTo(23.79, 2)
})

test('spargeVol', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).spargeVol
  ).toBeCloseTo(16.36, 2)
})

test('estPreBoilVolume', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).estPreBoilVolume
  ).toBeCloseTo(37.12, 2)
})

test('boilOffVolume', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).boilOffVolume
  ).toBeCloseTo(10.18, 2)
})

test('postBoilVolume', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).postBoilVolume
  ).toBeCloseTo(26.94, 2)
})

test('coolingShrinkage', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).coolingShrinkage
  ).toBeCloseTo(1.08, 2)
})

test('estBottlingVol', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).estBottlingVol
  ).toBeCloseTo(21.32, 2)
})

test('totalWater', () => {
  expect(
    calculateVolumes(AussieAle, AussieAle.equipment).totalWater
  ).toBeCloseTo(40.81, 2)
})
