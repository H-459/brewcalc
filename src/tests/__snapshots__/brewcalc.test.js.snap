// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`srm2css 1`] = `"rgb(156, 21, 0)"`;
